### ISSUE TITLE
Remember me doesn't work

### ISSUE THREAD

**Issue by DCKT** [heartsentwined/ember-auth#9]
Hi, 

I didn't saw, but the remember me functionality doesn't work, in the Chrome Web Tools we can see the cookie but when I refresh the page, the view return to sign in. 
Is it a bug of ember-auth or in my configuration ? 

```
Auth.Config.reopen
  tokenCreateUrl: '/api/login'
  tokenDestroyUrl: '/api/logout'
  tokenKey: 'auth_token'
  idKey: 'user_id'
  rememberMe: true
  rememberTokenKey: 'remember_token'
  rememberPeriod: 7
```

Thanks a lot

**Comment by heartsentwined** [comment]
You need to call `Auth.Module.RememberMe.recall()` explicitly when you want to restore the session. This example restores the session app-wide, once the `ApplicationView` has finished loading:

``` coffeescript
App.ApplicationView = Em.View.extend
  didInsertElement: ->
    Auth.Module.RememberMe.recall()
```

If this is not the problem, then please post your code for further debugging.

P.S. Why does it not happen automagically? The above example incurs overhead app-wide, and might not be the best fit for another app - say, for example, an app that's mostly public, except for a small auth-only area. In this case, it would be better to restore the session only in those auth-only areas.

**Comment by AshesOfOwls** [comment]
I bet I know what was wrong because I just ran into this issue.

If you're like me at all you tend to skim through tutorials trying to copy and paste as fast as possible, but that always leads to some minor mistakes. I was following your Rails integration tutorial and it seems I completely forgot to add the "remember" field to the sign_in view. I didn't think this was an issue because I figured that the options passed to Auth.config.reopen were default, and so I thought that I wouldn't need to include the remember field. By the time this issue came up I completely forgot about it being added to the sign_in view and got quite confused why it wasn't working.

So yeah, last thing to check is that you actually include remember: true in the sign_in view so that you actually have a cookie to remember you by.

Either way thanks a lot for this auth system and the rails integration!

**Comment by baraka2000** [comment]
Would it be terribly annoying to ask a beginner question. I managed to get
to the auth UI step, but my posts are not listing (nor are my users, but
login/out IS working)... I know the routes (and I'm assuming controllers)
are working because I can put random html into the post template and see it
in the browser. So things seemed to be working, just not loading my post
data....for some reason?? I have confirmed the data is in the db, and the
user model appears to be working (because I can login), my posts are setup
the same...

? thanks in advance.

**Comment by heartsentwined** [comment]
Thanks @McPants , you know I had your exact problem when I was skimming through my own tutorial, copy-and-pasting, to test that it works. Took me full 15 mins to figure out what's wrong :-)

**Comment by heartsentwined** [comment]
@baraka2000 No problem, everybody has been a beginner once. Here's what I would do for debugging, along with example errors that had happened to me somewhere along my own learning journey:
1. Normally you can't do this, but this is a tutorial with known working sample code, so just clone the `ember-auth-rails-demo` repo, and compare diff (I use [Meld](http://meldmerge.org/))
2. Real debugging starts. Open Firebug / Console, and examine the XHR requests being sent:
3. Are they being sent to the correct API end points? (e.g. missing the `/api` namespace, for example?)
4. Do the responses contain the expected data? (if not, you might hav "confirmed" that the data is in `development` db, but you are running a `production` server)
5. Are the data well-formed? (missing the JSON root wrapper? singular vs plural?)
6. Now you should know whether the problem is in client-side ember code, or server-side rails code. Here the possibilities are too many to be listed, but a couple of ideas:
7. Ember: Correct use (and omission) of trailing `()` in coffeescript method calls? Correct `model` setting for controllers? Handlebars / Hamlbars / etc syntax (you mentioned "plain HTML" is working)?
8. Rails: Correct serializer settings? Are you allowing JSON API access? (you should be) Anything in the server logs?

**Comment by baraka2000** [comment]
Thanks for your reply!

I realize my question was sort of like "it's not working?" Which is hard
for anyone to answer, but I appreciate your pointers. I have spent a lot of
time looking at your code on github, "manual diff'ing" :) I'll try to get
meld working. The thing is I tried to be adventurous and setup a slightly
custom model instead of posts, thinking I might later try to us this sample
app. ASide from that I'm pretty sure I've ma

**Comment by baraka2000** [comment]
Thanks for your reply! I realize my question is a bit unspecific, like
"it's not working?" So I appreciate your pointers.

Looking at the XHR log in the console there are no errors, but closely
examining it I don't see any data in there? I checked my production db, it
was empty, so I migrated the production db thinking this would fix my
problem, and still no luck... Checked server logs, no errors there, but I
don't see anything that looks like JSON coming in. The strange thing is
signing in works, locally, but not on my heroku deployed version (have run
migrations and seeding on heroku). I think I have everything wired up
correctly (since straight html displays from my posts template), and I have
"manually diff'd" for hours looking over your code on github, everything
appears correct. Specs all pass.

Couple of questions:

1) How do I know which server I'm running? prod versus dev using this
foreman/Procfile/unicorn config? Is this always prod?
2) How do I allow JSON API access? I thought this was just through the
render in the rails controller, is there some config I missed?

This is my heroku version: http://infinite-everglades-2103.herokuapp.com/

Note I was adventurous and changed the name of the posts model to "proofs."
But I think I made all the naming changes throughout the tutorial as
necessary.

Thanks for your help.
Kevin

**Comment by heartsentwined** [comment]
A couple of things:
1. I haven't dug deep into your ember side code yet, but the first sign of error is the XHR requests. It contains no body. A `curl -v http://infinite-everglades-2103.herokuapp.com/api/proofs` gives a `200 OK`, but it renders the standard Rails layout (instead of a JSON). `curl -v http://infinite-everglades-2103.herokuapp.com/api/proofs.json` (note the `.json`) gives a `500 Intenal Server Error`.
2. There is no debug message, so you are running a production server on heroku. (Good)
3. You should look into server logs (`heroku logs`) to see what happened, but I would guess you did not allow JSON access to API calls, and consequently you did not _limit_ API end point to JSON access only.
4. In any case, your tests should have failed long ago, because your `/api/proofs` API end point is not working. (see below if this is not the case)

At this point, I suggest that you re-do the [Model](https://github.com/heartsentwined/ember-auth-rails-demo/wiki/Model) section of the tutorial. The tests should have been failing already; there's no harm renaming all `posts` into `proofs`.

About 1/3 into the page there is the code for enable JSON access to API end points (and limit them to JSON only). Here's the code:

``` rb
module Api
  class BaseController < ApplicationController
    # ...
    respond_to :json
    before_filter :default_json

    protected

    def default_json
      request.format = :json if params[:format].nil?
    end
  end
end
```

If tests are passing in dev environment, then something is wrong with production environment settings. Most likely it would be in `config/environments/production.rb`, although it could also be the case that something in `config/environments/development.rb` is _masking_ over the problem at dev. In this case, I would compare the settings, and if I still have no idea which setting is wrong, try googling or posting those settings here. Or brute-force approach. (start from dev setting, change them one by one to production setting, see what caused the error)

As to your other question, you can pass the environment variable `RAILS_ENV` to foreman. For example:

``` sh
RAILS_ENV=production foreman start
```

will force foreman to use the production environment on your local machine.

You can also create a file called `.env` at your rails project root and put `RAILS_ENV=production` in it for the same effect.

**Comment by baraka2000** [comment]
I just navigated to
http://infinite-everglades-2103.herokuapp.com/api/proofs and
get

Error: No route matched the URL '/api/proofs'

pointing to routes, I just don't know how to fix. When I "manually
diff" all looks the same as your routes setup minus the name change...

**Comment by baraka2000** [comment]
Fixed it! It was my routes.rb

the order of my routes was eff'ing it up, I had:

  root to: 'home#index'

  get '*ember' => 'home#index'

  devise_for :users, controllers: {sessions: 'sessions'}

  namespace :api do
    resources :proofs, only: [:index, :show]
    resources :users, only: [:index, :show]
  end

instead of:

  namespace :api do
    resources :proofs, only: [:index, :show]
    resources :users, only: [:index, :show]
  end

  devise_for :users, controllers: {sessions: 'sessions'}

  root to: 'home#index'

  get '*ember' => 'home#index'

THANK YOU!

**Comment by heartsentwined** [comment]
Glad you found the solution.

@DCKT has your issue been resolved yet? Hopefully the silence means all's going well; closing issue now.

**Comment by DCKT** [comment]
@heartsentwined Sorry for this late response. I remade the login by myself and it works ! But I've got a strange problem. 

When I check for cookies and one found, the login view is loaded for 1 sec and immediatly change to the main view. 
If someone had already this case can help me. Thanks a lot.

EDIT : I found how to correct that, we just have to block our ajax request so `async: false`

**Comment by heartsentwined** [comment]
@DCKT this is an old thread but the latest version has fixed #8 which is, as far as I can tell, the problem that you have described here.

If remember me is enabled, the `ember-auth` will do a blocking ajax request (like your `async: false`) only on initial page load, before the redirection, so it should not flicker for 1 sec or so to the login view if there is a valid remembered session via cookie.
